redis/asyncio/cluster.py

```diff
@@ -0,0 +1,1118 @@
+                # self-healed, we will try to reinitialize the cluster layout
+                # and retry executing the command
+                await asyncio.sleep(0.25)
+                await self.initialize(force=True)
+                raise e
+            except ResponseError as e:
+                message = e.__str__()
+                log.exception(f"ResponseError: {message}")
+                raise e
+            except BaseException as e:
+                log.exception("BaseException")
+                if connection:
+                    await connection.disconnect()
+                raise e
+            finally:
+                if connection is not None:
+                    await redis_node.connection_pool.release(connection)
+
+        raise ClusterError("TTL exhausted.")
+
+    async def close(self):
+        try:
+            async with self._lock:
+                if self.nodes_manager:
+                    await self.nodes_manager.close()
+        except AttributeError:
+            # RedisCluster's __init__ can fail before nodes_manager is set
+            pass
+
+    def _process_result(self, command, res, **kwargs):
+        """
+        Process the result of the executed command.
+        The function would return a dict or a single value.
+
+        :type command: str
+        :type res: dict
+
+        `res` should be in the following format:
+            Dict<node_name, command_result>
+        """
+        if command in self.result_callbacks:
+            return self.result_callbacks[command](command, res, **kwargs)
+        elif len(res) == 1:
+            # When we execute the command on a single node, we can
+            # remove the dictionary and return a single response
+            return list(res.values())[0]
+        else:
+            return res
+
+
+class ClusterNode:
+    def __init__(self, host, port, server_type=None, redis_connection=None):
+        if host == "localhost":
+            host = socket.gethostbyname(host)
+
+        self.host = host
+        self.port = port
+        self.name = get_node_name(host, port)
+        self.server_type = server_type
+        self.redis_connection = redis_connection
+
+    def __repr__(self):
+        return (
+            f"[host={self.host},"
+            f"port={self.port},"
+            f"name={self.name},"
+            f"server_type={self.server_type},"
+            f"redis_connection={self.redis_connection}]"
+        )
+
+    def __eq__(self, obj):
+        return isinstance(obj, ClusterNode) and obj.name == self.name
+
+    def __del__(self):
+        try:
+            if self.redis_connection is not None:
+                loop = asyncio.get_event_loop()
+                coro = self.redis_connection.close(True)
+                if loop.is_running():
+                    loop.create_task(coro)
+                else:
+                    loop.run_until_complete(coro)
+        except Exception:
+            pass
```

Comment:
@Andrew-Chen-Wang How should deletion for `ClusterNode` be handled?

Reply:
correct me if I'm wrong, but on Cluster creation, these ClusterNodes are always with that one Cluster object, as in there is no dynamic changes to these ClusterNodes, right? If so, you can take a look at how I handle connection pools in async Redis's client class's close method that can be replicated in Cluster's close method. 

I have also not take a good look at the code yet, but this is just my initial thought

Reply:
No, that's the problem. In case of `MOVED` errors or persistent connection errors, the nodes_manager would be reinitialized, which could lead to some Nodes being removed.